packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+
+```js
+constructor() {
+  const defaults = {setting1: 'value1'};
+  this.settings = new DeviceSettings(defaults, this.onSettingsUpdate.bind(this));
+}
+
+async onSettingsUpdate(key, value) {
+  // do anything you want here with key and value
+}
+```
+
+### Make itself aware of resources other concurrent drivers are using
+
+Let's say your driver uses up some system resources, like ports. There are a few ways to make sure
+that multiple simultaneous sessions don't use the same resources:
+
+1. Have your users specify resource IDs via capabilities (`appium:driverPort` etc)
+1. Just always use free resources (find a new random port for each session)
+1. Have each driver express what resources it is using, then examine currently-used resources from
+   other drivers when a new session begins.
+
+To support this third strategy, you can implement `get driverData` in your driver to return what
+sorts of resources your driver is currently using, for example:
+
+```
+get driverData() {
+  return {specialPort: 1234, specialFile: /path/to/file}
+}
+```
+
+Now, when a new session is started on your driver, the `driverData` response from any other
+simultaneously running drivers will also be included, as the last parameter of the `createSession`
+method:
+
+```
+async createSession(jwpCaps, reqCaps, w3cCaps, driverData)
+```
+
+You can dig into this `driverData` array to see what resources other drivers are using to help
+determine which ones you want to use for this particular session.
+
+### Log events to the Appium event timeline
+
+Appium has an [Event Timing API](../guides/event-timing.md) which allows users to get timestamps for
+certain server-side events (like commands, startup milestones, etc...) and display them on
+a timeline. You can add your own events to the event log:
+
+```js
+this.logEvent(name);
+```
+
+Simply provide a name for the event and it will be added at the current time, and made accessible
+as part of the event log for users.
+
+### Hide behaviour behind security flags
+
+Appium has a feature-flag based [security model](../guides/security.md) that allows driver authors
+to hide certain features behind security flags. What this means is that if you have a feature you
+deem insecure and want to require server admins to opt in to it, you can require that they enable
+the feature by adding it to the `--allow-insecure` list or turning off server security entirely.
+
+To support the check within your own driver, you can call `this.isFeatureEnabled(featureName)` to
+determine whether a feature of the given name has been enabled. Or, if you want to simply
+short-circuit and throw an error if the feature isn't enabled, you can call
+`this.ensureFeatureEnabled(featureName)`.
+
+### Use a temp dir for files
+
+If you want to use a temporary directory for files your driver creates that are not important to
+keep around between computer or server restarts, you can simply read from `this.opts.tmpDir`.
```

Comment:
Is this a string? does it reference a temp dir that already exists? how can the user control when it's cleaned up? what is an example use-case?

Reply:
the idea behind this is that one could provide a path to this dir from a server command line arg and then clean it safely up from the client code after the server process gets killed.

Reply:
this is also the main reason to prefer using tmpDir module from `@appium/support` instead of directly using ones provided by Node.